src/failure_detector.rs

```diff
@@ -0,0 +1,373 @@
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::{HashMap, HashSet};
+use std::sync::RwLock;
+use std::time::Duration;
+#[cfg(not(test))]
+use std::time::Instant;
+
+#[cfg(test)]
+use mock_instant::Instant;
+use serde::{Deserialize, Serialize};
+use tracing::debug;
+
+/// A phi accrual failure detector implementation.
+pub struct FailureDetector {
+    /// Heartbeat samples for each node.
+    node_samples: RwLock<HashMap<String, SamplingWindow>>,
+    /// Failure detector configuration.
+    config: FailureDetectorConfig,
+    /// Denotes live nodes
+    live_nodes: HashSet<String>,
+    /// Denotes dead nodes
+    dead_nodes: HashSet<String>,
+}
+
+impl FailureDetector {
+    pub fn new(config: FailureDetectorConfig) -> Self {
+        Self {
+            node_samples: RwLock::new(HashMap::new()),
+            config,
+            live_nodes: HashSet::new(),
+            dead_nodes: HashSet::new(),
+        }
+    }
+
+    /// Report node heartbeat
+    pub fn report_heartbeat(&mut self, node_id: &str) {
+        debug!(node_id = node_id, "reporting node heartbeat.");
+        let mut node_samples = self.node_samples.write().unwrap();
+        let heartbeat_window = node_samples.entry(node_id.to_string()).or_insert_with(|| {
+            SamplingWindow::new(
+                self.config.window_size,
+                self.config.max_interval,
+                self.config.initial_interval,
+            )
+        });
+        heartbeat_window.report_heartbeat();
+    }
+
+    /// Marks/unmarks node as dead
+    pub fn update_node_liveliness(&mut self, node_id: &str) {
+        let phi = self.phi(node_id);
+        debug!(node_id = node_id, phi = phi, "updating node liveliness");
+        if phi > self.config.phi_treshold {
+            self.live_nodes.remove(node_id);
+            self.dead_nodes.insert(node_id.to_string());
+            // Remove current sampling window so that when the node
+            // comes back online, we start with a fresh sampling window.
+            self.remove(node_id);
+        } else {
+            self.live_nodes.insert(node_id.to_string());
+            self.dead_nodes.remove(node_id);
+        }
+    }
+
+    /// Returns live nodes
+    pub fn live_nodes(&self) -> impl Iterator<Item = &str> {
+        self.live_nodes.iter().map(|node_id| node_id.as_str())
+    }
+
+    /// Returns dead nodes
+    pub fn dead_nodes(&self) -> impl Iterator<Item = &str> {
+        self.dead_nodes.iter().map(|node_id| node_id.as_str())
+    }
+
+    /// Removes a node.
+    fn remove(&mut self, node_id: &str) {
```

Comment:
When is this API called? Shall we also remove the node from the live or dead node list?

Reply:
It's called on https://github.com/quickwit-oss/scuttlebutt/pull/13/files#diff-bbdade1c6c33834bc204bfb0012ca6a1eadf0c6ff3fc612484dc837f5496648dR74-R76
I preferred being explicit and decouple from live/dead nodes management.

Reply:
Thanks. Missed that.